[CIR] Implement cc1 support for a CIR pipeline

lanza opened the pull request:
This patch implements trivial cc1 support that'll let you pass
    `-fenable-clangir` to run clang with the clangir pipeline inserted. This
    also adds `-emit-cir` which will output a cir file and `-emit-cir-only`
    which will run the pipeline stopping after CIRGen.

    $ clang -cc1 test.c -fenable-clangir -emit-cir-only        (no output)
    $ clang -cc1 test.c -fenable-clangir -emit-cir             (outputs test.cir)
    $ clang -cc1 test.c -fenable-clangir -emit-cir -o here.cir (outputs here.cir)

    The following invocation:

    $ clang -cc1 test.c -fenable-clangir

    will crash. But it will eventually work as stanard clang does honoring
    your standard permutations of `-emit-llvm`, `-S`, `-c`,
    `-emit-llvm-only` etc.

bcardosolopes commented on clang/lib/Driver/Types.cpp:
```diff
@@ -387,9 +387,17 @@ types::getCompilationPhases(const clang::driver::Driver &Driver,
+  // something...
+  else if (DAL.getLastArg(options::OPT_emit_llvm) &&
+           DAL.getLastArg(options::OPT_fuse_cir))
+    LastPhase = phases::CIR;
```
I think it makes sense if you pass both `-emit-llvm -fuse-cir` that it will emit LLVM out of CIR. That doesn't mean we need to support it right now though. What happens if both `-emit-llvm` and `-emit-cir` are both used?

bcardosolopes replied:
Cool! Can you write down the current behavior in `CIR.rst`? It will probably be helpful to guide us when we forget hehe!

bcardosolopes replied:
Yea, until we have a need for this I'd think it's not worth working on it. For the rest, I'm fine with any model for now and it would be good to have some driver level for them, don't worry too much, we can always change later.

bcardosolopes replied:
No need to break it further, free feel to push the version you feel more like doing now!

bcardosolopes commented on clang/include/clang/Driver/Options.td:
```diff
@@ -749,6 +749,8 @@ def R_Joined : Joined<["-"], "R">, Group<R_Group>, Flags<[CC1Option, CoreOption]
   MetaVarName<"<remark>">, HelpText<"Enable the specified remark">;
 def S : Flag<["-"], "S">, Flags<[NoXarchOption,CC1Option]>, Group<Action_Group>,
   HelpText<"Only run preprocess and compilation steps">;
+def fuse_cir : Flag<["-"], "fuse-cir">, Flags<[CC1Option]>, Group<Action_Group>,
```
I think this name is a bit confusing, `-fcir-codegen` or similar would be a better description of what it does.

bcardosolopes replied:
That's a big name! How about `-fcir-to-llvm`?

lanza replied:
> Thinking about this some more, it seems odd to have both -fuse-cir and -emit_cir, why don't we only use -emit-cir which will have the accumulated meaning: Use the new ClangIR pipeline to compile + Build ASTs and then lower to ClangIR, and there continues the pipeline to LLVM and whatnots. Emission of the .cir file to disk is controlled by -fcir-output. If we want to stop after CIR generation, we then use -emit-cir-only, as a reference to -fsyntax-only and other flags.

That's a valid alternative. `-fsyntax-only`-like pattern to match the rest of the frontend instead of a `-emit-xyz` to match the backend. I guess being the new "middle-end" makes that choice non-obvious. I guess the differentiator here would be whether or not we think the flags should be exposed to the driver or only cc1. `-fsyntax-only` family of flags are cc1 only, `-emit-xyx` flags are handled by the driver as well.

bcardosolopes replied:
I think we have some freedom. I mentioned `-fsyntax-only` mostly for the "only" part. There's also `-emit-codegen-only`, `-emit-llvm-only` as other examples!

bcardosolopes commented on clang/include/clang/Driver/Types.def:
```diff
@@ -35,25 +35,25 @@
-TYPE("objc++-cpp-output",        PP_ObjCXX_Alias, INVALID,      "mii",    phases::Compile, phases::Backend, phases::Assemble, phases::Link)
-TYPE("objective-c++",            ObjCXX,       PP_ObjCXX,       "mm",     phases::Preprocess, phases::Compile, phases::Backend, phases::Assemble, phases::Link)
-TYPE("renderscript",             RenderScript, PP_C,            "rs",     phases::Preprocess, phases::Compile, phases::Backend, phases::Assemble, phases::Link)
+TYPE("cpp-output",               PP_C,         INVALID,         "i",      phases::Compile, phases::CIR, phases::Backend, phases::Assemble, phases::Link)
```
Is it possible that CIR stuff is hidden behind/inside `phases::Backend`?

lanza replied:
I was thinking I could leave it in the `compile` phase, but the extra logic required there would be pretty cumbersome as well. I'm not really happy with either idea. Do you have a reasonably strong opinion about whether or not this is right?

bcardosolopes replied:
What do you see changing by adding something in this table? Do you have a testcase or something that breaks? I'd suggest leaving this out for now if it's not providing any real change needed by other parts.

bcardosolopes reviewed (approved):
This is awesome Nathan. I'm gonna approve this but it's up to you to iterate over the comments now or as part of a new PR.

lanza commented:
Okay, I iterated through it some more. The bikeshed looks better in blue 😀

* I implemented the forwarding of the `-o` through the `CIRGenAction` so that the `-fcir-output` flag is redundant. But you might want it to stay around for usage from the `CIRWarnings` side? Up to you there. I removed it since it become redundant for the compilation pipeline but can bring it back easy. 
* I added the `-emit-cir-only` flag. It behaves the same as the rest of the `only`s.
* I dropped the driver code. It's a bit muddier on what that should look like so might as well ignore it for now. We can just live in `-cc1`. 
* I switched the pipeline flag to `-fenable-clangir`. I asked a couple people what they thought the flag that toggled clangir should be and they both said `enable` and was persuaded.
* Compilation crashes if you attempt to invoke some unsupported pipeline. e.g. `$ clang -cc1 -fenable-clangir -emit-llvm test.c`. Once I get the lowering patch implemented we can enable the two `llvm` flavors. 
* `CIRContext` being an `ASTConsumer` is pretty weird. Maybe making the `CIRBuilder` the consumer and giving it a handle on the `CIRContext`?

The current flag setup looks like this:

```
$ clang -cc1 -fenable-clangir test.c -emit-cir           (emits test.cir)
$ clang -cc1 -fenable-clangir test.c -emit-cir-only      (emits nothing)
$ clang -cc1 -fenable-clangir test.c                     (crashes, NYI)
$ clang -cc1 test.c -emit-cir                            (crashes, nonsense to ask for CIR out of non-CIR pipeline)
```

You mentioned the `-fenable-clangir test.c -emit-cir-only -fcir-output=test.cir` which is covered by the normal `-o` behavior now, though we can still bring back `-fcir-output` if you'd like.

bcardosolopes commented:
>  But you might want it to stay around for usage from the CIRWarnings side?

This would be good for now, thanks!

> CIRContext being an ASTConsumer is pretty weird. Maybe making the CIRBuilder the consumer and giving it a handle on the CIRContext

Makes total sense.

Overall LGTM
